### PR TITLE
MSP

### PR DESCRIPTION
**Pull request by abdel**
- Added flag to build PM or CA1/CA3_CA1 pathway
- Introduced two new networks:
    - CA1 (auto-encoder on ECin) - continuously trains without resetting, bigger, slower lr
    - CA3_CA1 (reproducing CA1 hidden, from CA3 pattern) - resets alongside PR, fast lr
- During recall, the CA1 reconstruction of EC will be driven by CA3 (configurable by param, on by default)
- Added decoding through layers to reconstruct the CA1 <=> ECout back into pixels (through EC => LTM)
- Introduced unpooling in the VC to facilitate decoding backwards through layers

See also diary for more notes on issues/debugging.

https://docs.google.com/document/u/1/d/1oI_VNPB-ldBKQZ4dT5dWJF5ifjh9NrCmKqVv0lggIMA/edit?usp=drive_web&ouid=101721042953729020097

### CODE REVIEW

File: cls_module/cls_module/cls.py [hunk 1]
```diff
@@ -248,13 +248,24 @@ def forward(self, inputs, labels=None, mode=None):  # pylint: disable=arguments-
...
       if self.ec_key in self._modules:
-        ec = self._modules[self.ec_key]
         outputs[self.ec_key] = self.ec(inputs=next_input)
         next_input = outputs[self.ec_key].detach()  # Ensures no gradients pass through modules
 
       # iterate STM
       losses[self.stm_key], outputs[self.stm_key] = self.stm(inputs=next_input, targets=inputs, labels=labels)
 
+      # Decode output from STM via the EC <=> LTM
+      if outputs[self.stm_key]['memory']['decoding'] is None:
```
Comment (affogato) [inline cls_module/cls_module/cls.py:258]: Why does it check here? 
Is there a possibility that it was set already, and that we don't want to overwrite it?

File: cls_module/cls_module/cls.py [hunk 2]
```diff
@@ -248,13 +248,24 @@ def forward(self, inputs, labels=None, mode=None):  # pylint: disable=arguments-
...
+        output_decoding = outputs[self.stm_key]['memory']['decoding_ec']
+
+        if self.ec_key in self._modules:
+          output_decoding = self.ec.forward_decode(output_decoding)
+
+        output_decoding = self.ltm.forward_decode(output_decoding)
```
Comment (affogato) [inline cls_module/cls_module/cls.py:264]: Does this work for our AE LTM as well as the VGG LTM ?

File: cls_module/cls_module/memory/stm/aha.py [hunk 3]
```diff
@@ -185,15 +193,27 @@ def build(self):
...
+                    input_shape=self.input_shape,
+                    target_shape=self.input_shape)
+
+      ca1_output_shape = [1, self.config['ca1']['num_units']]
 
-    # Build the Pattern Mapping (PM) module - to an arbitrary 'target' (main use-case is the original input image)
-    self.build_ae(name='pm', config=self.config['pm'],
-                  input_shape=pc_output_shape,
-                  target_shape=self.target_shape)
+      self.build_ae(name='ca3_ca1', config=self.config['ca3_ca1'],
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha.py:204]: AE is convenient. Conceptually I think of it as FCNN, with the labels being CA1 hidden states. Is it equivalent to a 2 layer FCNN?

File: cls_module/cls_module/memory/stm/aha.py [hunk 4]
```diff
@@ -458,17 +478,44 @@ def forward_memory(self, inputs, targets, labels):
...
 
-    losses['pm'], outputs['pm'] = self.forward_ae(name='pm', inputs=outputs['pc'], targets=targets)
-    losses['pm_ec'], outputs['pm_ec'] = self.forward_ae(name='pm_ec', inputs=outputs['pc'], targets=targets)
+    if self.config.get('msp_type', None) == 'ca1':
+      # During study, EC will drive the CA1
+      losses['ca1'], outputs['ca1'] = self.forward_ae(name='ca1', inputs=inputs, targets=inputs)
+      ca3_ca1_target = outputs['ca1']['encoding']
+
+      losses['ca3_ca1'], outputs['ca3_ca1'] = self.forward_ae(name='ca3_ca1',
+                                                              inputs=outputs['pc'],
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha.py:487]: It would be more biological to use CA3 out, rather than PC.
Yes, it is the same in this case, and this is more straightforward.
But we should not forget about this.

File: cls_module/cls_module/memory/stm/aha.py [hunk 5]
```diff
@@ -458,17 +478,44 @@ def forward_memory(self, inputs, targets, labels):
...
     outputs['pc'] = self.forward_pc(inputs=pc_cue)
 
-    losses['pm'], outputs['pm'] = self.forward_ae(name='pm', inputs=outputs['pc'], targets=targets)
-    losses['pm_ec'], outputs['pm_ec'] = self.forward_ae(name='pm_ec', inputs=outputs['pc'], targets=targets)
+    if self.config.get('msp_type', None) == 'ca1':
+      # During study, EC will drive the CA1
+      losses['ca1'], outputs['ca1'] = self.forward_ae(name='ca1', inputs=inputs, targets=inputs)
+      ca3_ca1_target = outputs['ca1']['encoding']
+
+      losses['ca3_ca1'], outputs['ca3_ca1'] = self.forward_ae(name='ca3_ca1',
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha.py:486]: Do you think it would be cleaner code to have a reference to the actual AE, rather than using forward_ae?
Then we could change the implementation in future, e.g. to something that is not an ae, and it would still work.

Comment (affogato) [inline cls_module/cls_module/memory/stm/aha.py:486]: This is NOT within a conditional to ensure that it is in 'study'.
When in 'recall', this CA1 forward pass will be redundant.

Also, I thought it should fully train CA1 on all the current stimuli, and then train CA3-CA1. Is that how this will work? i.e. accumulate gradients across the batch, update weights, then we have a stable CA1 encoding to learn to reproduce by CA3-CA1. When hebbian, interleaved like this will work, because they'll both converge. But with batch training, I'm not sure.

File: lake/oneshot_cls.py [hunk 6]
```diff
@@ -263,7 +263,12 @@ def main():
     recall_target = torch.from_numpy(np.array(recall_target)).to(device)
 
     # Reset to saved model
-    model.load_state_dict(torch.load(pretrained_model_path))
+    model_dict = model.state_dict()
+    pretrained_dict = torch.load(pretrained_model_path)
+    pretrained_dict = {k: v for k, v in pretrained_dict.items() if k in model_dict}
+    model_dict.update(pretrained_dict)
```
Comment (affogato) [inline lake/oneshot_cls.py:269]: this line looks redundant?

File: lake/oneshot_cls.py [hunk 7]
```diff
@@ -263,7 +263,12 @@ def main():
     recall_target = torch.from_numpy(np.array(recall_target)).to(device)
 
     # Reset to saved model
-    model.load_state_dict(torch.load(pretrained_model_path))
+    model_dict = model.state_dict()
+    pretrained_dict = torch.load(pretrained_model_path)
+    pretrained_dict = {k: v for k, v in pretrained_dict.items() if k in model_dict}
```
Comment (affogato) [inline lake/oneshot_cls.py:268]: why is this necessary?

File: cls_module/cls_module/memory/stm/aha.py (hunk 3 above)
Comment (abdel) [reply cls_module/cls_module/memory/stm/aha.py:204]: Since the AE used is a 'simple AE', which is just two FC layers. it is essentially the same thing as a 2 layer FCNN.

The downside is that the AE assumes the number of units on the two layers are equal.

File: cls_module/cls_module/cls.py (hunk 2 above)
Comment (abdel) [reply cls_module/cls_module/cls.py:264]: It works for the AE LTM aka VC. I have not yet tested with VGG LTM though.. but I can

File: cls_module/cls_module/cls.py (hunk 1 above)
Comment (abdel) [reply cls_module/cls_module/cls.py:258]: If it was already set, then it means that we had some sort of PM-like network that reproduced the 'raw' image and set this variable. If not, it means we are either using CA1 or purely PM_EC.

File: lake/oneshot_cls.py (hunk 7 above)
Comment (abdel) [reply lake/oneshot_cls.py:268]: Optional to keep it, but it was motivated by the fact that I wanted to use a pretrained LTM (that had the PM in it) but PyTorch refuses to load it because I hadn't built the PM.

This line of code essentially only loads the parts that you actually initalised by checking against the instantiated model and comparing it with what's being loaded from the pretrained path.

File: cls_module/cls_module/memory/stm/aha.py (hunk 5 above)
Comment (abdel) [reply cls_module/cls_module/memory/stm/aha.py:486]: Yes, the forward pass will be redundant with regards to the final output as CA1 will be driven by CA3. I kept it because it can provide a target for CA3 during recall to provide a loss, otherwise we would have no target during recall to compare to.

Maybe the 'forward' naming is misleading, but each of our components' "forward_XX" method does both a forward pass and a backward pass (if training flag is on).

So in this case, the `forward_ae` with `ca1` has done a complete optimization step on the current batch, before moving onto the forward_ae with `ca3_ca1`

Comment (abdel) [reply cls_module/cls_module/memory/stm/aha.py:486]: Do you mean similar to the PR/PS/PC forward methods?

File: cls_module/cls_module/cls.py (hunk 2 above)
Comment (affogato) [reply cls_module/cls_module/cls.py:264]: But can VGG decode? i.e. not not AE

File: cls_module/cls_module/memory/stm/aha.py (hunk 3 above)
Comment (affogato) [reply cls_module/cls_module/memory/stm/aha.py:204]: That seems like a big downside. Do we have a re-usable component that is a 2 layer FCNN. Any reason not to use it?

File: cls_module/cls_module/memory/stm/aha.py (hunk 5 above)
Comment (affogato) [reply cls_module/cls_module/memory/stm/aha.py:486]: Yeah, something like self.
`self.ca1.forward()`

File: cls_module/cls_module/memory/stm/aha.py (hunk 3 above)
Comment (abdel) [reply cls_module/cls_module/memory/stm/aha.py:204]: No, I think at that point we should just construct CA1, etc. as a separate module, but with torch.nn.Linear layers

File: lake/oneshot_cls.py (hunk 6 above)
Comment (affogato) [reply lake/oneshot_cls.py:269]: Any thoughts on this one?

File: cls_module/cls_module/cls.py (hunk 2 above)
Comment (abdel) [reply cls_module/cls_module/cls.py:264]: yeah, this won't work with current VGG actually. Not without adding a separate decoder layer.

File: lake/oneshot_cls.py (hunk 6 above)
Comment (abdel) [reply lake/oneshot_cls.py:269]: I am going to assume you are commented about `model_dict.update(pretrained_dict)` - if so, then yes it is redundant! 

I was confused by the GitHub showing the same code snippet but with different comments

File: cls_module/cls_module/memory/stm/aha.py (hunk 5 above)
Comment (abdel) [reply cls_module/cls_module/memory/stm/aha.py:486]: Yeah, as I have been adding more sub-modules.. I have been feeling that it is time to actually split them out, rather than keep all the builds/initialisations & forwards all inside the single AHA module

Comment (abdel) [reply cls_module/cls_module/memory/stm/aha.py:486]: We may need to discuss how we want to split/organise them though
